File: ivc-client/src/main/java/ivc/IvcClient.java
package ivc;

import java.io.IOException;
import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.CompletableFuture;

/**
 * IvcClient — HTTP transport layer for the IVC API (Java 17 / JDK HttpClient).
 *
 * Configuration:
 *   IvcClient.configure("https://my.ivc.host");
 *
 * Defaults (in order):
 *   1. Last value passed to configure()
 *   2. System property ivc.base.url
 *   3. Environment variable IVC_BASE_URL
 *   4. "" (relative — caller must set explicitly)
 */
public final class IvcClient {

    private IvcClient() {}

    private static String _baseUrl = resolveDefaultBaseUrl();
    private static Duration _timeout = Duration.ofSeconds(10);

    private static final HttpClient HTTP = HttpClient.newBuilder()
            .version(HttpClient.Version.HTTP_1_1)
            .connectTimeout(Duration.ofSeconds(5))
            .build();

    // ---------------------------------------------------------------
    // Configuration
    // ---------------------------------------------------------------

    public static void configure(String baseUrl) {
        _baseUrl = baseUrl.replaceAll("/$", "");
    }

    public static void configure(String baseUrl, Duration timeout) {
        configure(baseUrl);
        _timeout = timeout;
    }

    private static String resolveDefaultBaseUrl() {
        String prop = System.getProperty("ivc.base.url");
        if (prop != null && !prop.isBlank()) return prop.replaceAll("/$", "");
        String env  = System.getenv("IVC_BASE_URL");
        if (env  != null && !env.isBlank())  return env.replaceAll("/$", "");
        return "";
    }

    // ---------------------------------------------------------------
    // Raw HTTP helpers
    // ---------------------------------------------------------------

    public static CompletableFuture<IvcRawResult> get(String path) {
        HttpRequest req = HttpRequest.newBuilder()
                .uri(URI.create(_baseUrl + path))
                .timeout(_timeout)
                .header("X-IVC-Client", "java/1.0")
                .GET()
                .build();
        return HTTP.sendAsync(req, HttpResponse.BodyHandlers.ofString())
                .thenApply(IvcClient::processResponse);
    }

    public static CompletableFuture<IvcRawResult> post(String path, String jsonBody) {
        HttpRequest req = HttpRequest.newBuilder()
                .uri(URI.create(_baseUrl + path))
                .timeout(_timeout)
                .header("Content-Type", "application/json")
                .header("X-IVC-Client", "java/1.0")
                .POST(HttpRequest.BodyPublishers.ofString(jsonBody, StandardCharsets.UTF_8))
                .build();
        return HTTP.sendAsync(req, HttpResponse.BodyHandlers.ofString())
                .thenApply(IvcClient::processResponse);
    }

    public static CompletableFuture<IvcRawResult> put(String path, String jsonBody) {
        HttpRequest req = HttpRequest.newBuilder()
                .uri(URI.create(_baseUrl + path))
                .timeout(_timeout)
                .header("Content-Type", "application/json")
                .header("X-IVC-Client", "java/1.0")
                .PUT(HttpRequest.BodyPublishers.ofString(jsonBody, StandardCharsets.UTF_8))
                .build();
        return HTTP.sendAsync(req, HttpResponse.BodyHandlers.ofString())
                .thenApply(IvcClient::processResponse);
    }

    private static IvcRawResult processResponse(HttpResponse<String> res) {
        String raw = res.body();
        Map<String, Object> body = JsonSimple.parse(raw);

        // Parse Status: header
        IvcStatus status = null;
        Optional<String> sh = res.headers().firstValue("Status");
        if (sh.isEmpty()) sh = res.headers().firstValue("X-IVC-Status");
        if (sh.isPresent()) status = IvcMarshaller.parseStatusHeader(sh.get());

        IrcObject object = IvcMarshaller.fromResponse(body);
        return new IvcRawResult(body, status, object);
    }

    // ---------------------------------------------------------------
    // IVC-specific helpers
    // ---------------------------------------------------------------

    /**
     * Fetch the current mode string for a target object from the server.
     * Endpoint: GET /api/object.php?target=<target>
     */
    public static CompletableFuture<IvcRawResult> fetchObject(String target) {
        String encoded = URLEncoder.encode(target, StandardCharsets.UTF_8);
        return get("/api/object.php?target=" + encoded);
    }

    /**
     * Apply a mode delta to a target object.
     * Endpoint: POST /api/modes.php
     */
    public static CompletableFuture<IvcResponse> applyModes(String target, String delta, String requester) {
        String json = String.format(
                "{\"target\":\"%s\",\"delta\":\"%s\",\"requester\":\"%s\"}",
                escape(target), escape(delta), escape(requester));
        return post("/api/modes.php", json).thenApply(r -> {
            boolean success = Boolean.TRUE.equals(r.body().get("success"));
            String  message = (String) r.body().getOrDefault("message", "");
            String  bt      = (String) r.body().get("base_target");
            String  modes   = (String) r.body().get("modes");
            IvcResponse res = new IvcResponse(success, message, bt, modes, r.status());
            return res;
        });
    }

    /**
     * Check access to a target.
     * Endpoint: GET /api/access.php?target=<uri>&user=<nick>
     */
    public static CompletableFuture<IvcRawResult> checkAccess(String target, String user) {
        String path = "/api/access.php?target=" + URLEncoder.encode(target, StandardCharsets.UTF_8);
        if (user != null && !user.isBlank()) {
            path += "&user=" + URLEncoder.encode(user, StandardCharsets.UTF_8);
        }
        return get(path);
    }

    private static String escape(String s) {
        return s == null ? "" : s.replace("\\", "\\\\").replace("\"", "\\\"");
    }

    // ---------------------------------------------------------------
    // Raw result container
    // ---------------------------------------------------------------

    public record IvcRawResult(Map<String, Object> body, IvcStatus status, IrcObject object) {}
}

File: src/IRC/IvcClient.ts
/**
 * IvcClient — HTTP transport layer for the IVC API.
 *
 * Wraps the Fetch API (browser or Node 18+) to:
 *   - Send requests to the IVC server with standard headers
 *   - Parse the `Status:` response header on every response
 *   - Invoke IvcMarshaller.fromResponse() to auto-hydrate typed objects
 *
 * Configuration:
 *   IvcClient.configure({ baseUrl: 'https://my.ivc.host' });
 *
 * Defaults (in order):
 *   1. Last value passed to configure()
 *   2. process.env.IVC_BASE_URL   (Node)
 *   3. window.location.origin     (browser)
 *   4. ''                         (relative URLs — same-origin)
 */

import { IvcMarshaller } from './IvcMarshaller.js';
import type { IvcApiResponse, IvcResponse, IvcStatus } from './types.js';

export interface IvcClientConfig {
  baseUrl?: string;
  /** Additional headers merged into every request. */
  headers?: Record<string, string>;
  /** Request timeout in milliseconds (default 10 000). */
  timeoutMs?: number;
}

interface IvcRawResult {
  body:   IvcApiResponse;
  status: IvcStatus | null;
  object: unknown | null;
}

let _config: IvcClientConfig = {};

export class IvcClient {
  // ---------------------------------------------------------------
  // Configuration
  // ---------------------------------------------------------------

  static configure(cfg: IvcClientConfig): void {
    _config = { ..._config, ...cfg };
  }

  static get baseUrl(): string {
    if (_config.baseUrl) return _config.baseUrl.replace(/\/$/, '');
    if (typeof process !== 'undefined' && process.env?.['IVC_BASE_URL']) {
      return (process.env['IVC_BASE_URL'] as string).replace(/\/$/, '');
    }
    if (typeof window !== 'undefined') return window.location.origin;
    return '';
  }

  // ---------------------------------------------------------------
  // Core request
  // ---------------------------------------------------------------

  private static async request(
    method:  'GET' | 'POST' | 'PUT' | 'DELETE',
    path:    string,
    body?:   Record<string, unknown>,
  ): Promise<IvcRawResult> {
    const url  = `${IvcClient.baseUrl}${path}`;
    const ctrl = new AbortController();
    const tid  = setTimeout(() => ctrl.abort(), _config.timeoutMs ?? 10_000);

    const headers: Record<string, string> = {
      'Content-Type': 'application/json',
      'X-IVC-Client':  'ts/1.0',
      ...(_config.headers ?? {}),
    };

    const init: RequestInit = {
      method,
      headers,
      signal: ctrl.signal,
    };
    if (body !== undefined) {
      init.body = JSON.stringify(body);
    }

    let res: Response;
    try {
      res = await fetch(url, init);
    } finally {
      clearTimeout(tid);
    }

    const responseBody = await res.json() as IvcApiResponse;

    // Parse Status: header
    let ivcStatus: IvcStatus | null = null;
    const statusHeader = res.headers.get('Status') ?? res.headers.get('X-IVC-Status');
    if (statusHeader) {
      ivcStatus = IvcMarshaller.parseStatusHeader(statusHeader);
    }

    // Auto-unmarshal
    const object = IvcMarshaller.fromResponse(responseBody);

    return { body: responseBody, status: ivcStatus, object };
  }

  // ---------------------------------------------------------------
  // HTTP verbs
  // ---------------------------------------------------------------

  static async get(path: string): Promise<IvcRawResult> {
    return IvcClient.request('GET', path);
  }

  static async post(path: string, body: Record<string, unknown>): Promise<IvcRawResult> {
    return IvcClient.request('POST', path, body);
  }

  static async put(path: string, body: Record<string, unknown>): Promise<IvcRawResult> {
    return IvcClient.request('PUT', path, body);
  }

  static async delete(path: string, body?: Record<string, unknown>): Promise<IvcRawResult> {
    return IvcClient.request('DELETE', path, body);
  }

  // ---------------------------------------------------------------
  // IVC-specific helpers
  // ---------------------------------------------------------------

  /**
   * Fetch the current mode string for a target object from the server.
   * Endpoint: GET /api/object.php?target=<target>
   */
  static async fetchObject(target: string): Promise<IvcRawResult> {
    const encoded = encodeURIComponent(target);
    return IvcClient.get(`/api/object.php?target=${encoded}`);
  }

  /**
   * Apply a mode delta to a target object.
   * Endpoint: POST /api/modes.php
   * Body: { target, delta, requester }
   */
  static async applyModes(
    target:    string,
    delta:     string,
    requester: string = '',
  ): Promise<IvcResponse> {
    const result = await IvcClient.post('/api/modes.php', { target, delta, requester });
    return {
      success:     result.body.success,
      message:     result.body.message ?? '',
      base_target: result.body.base_target,
      modes:       result.body.modes,
      object:      result.object as IvcResponse['object'],
      status:      result.status ?? undefined,
    };
  }

  /**
   * Check access to a target (mirrors ChanServ::checkAccess).
   * Endpoint: GET /api/access.php?target=<uri>&user=<nick>
   */
  static async checkAccess(target: string, user?: string): Promise<IvcRawResult> {
    let path = `/api/access.php?target=${encodeURIComponent(target)}`;
    if (user) path += `&user=${encodeURIComponent(user)}`;
    return IvcClient.get(path);
  }
}
